src/mongo/db/pipeline/expression.cpp

```diff
@@ -3900,6 +3935,65 @@ Value ExpressionSlice::evaluate(const Document& root) const {
     return Value(vector<Value>(array.begin() + start, array.begin() + end));
 }
 
+intrusive_ptr<Expression> ExpressionSlice::optimize() {
+    // If ExpressionSlice is passed an ExpressionFilter we can stop filtering once the size of
+    // the array returned by the filter is equal to the last arguement passed to ExpressionSlice.
+    if (dynamic_cast<ExpressionFilter*>(vpOperand[0].get())) {
+        if (auto secondArg = dynamic_cast<ExpressionConstant*>(vpOperand[1].get())) {
+            auto secondVal = secondArg->getValue();
+
+            uassert(50797,
+                    str::stream() << "Second argument to $slice must be a numeric value,"
+                                  << " but is of type: "
+                                  << typeName(secondVal.getType()),
+                    secondVal.numeric());
+
+            uassert(50798,
+                    str::stream() << "Second argument to $slice can't be represented as"
+                                  << " a 32-bit integer: " 
+                                  << secondVal.coerceToDouble(),
+                    secondVal.integral());
+
+            int arg2 = secondVal.coerceToInt();
+            if (vpOperand.size() == 2) {
+                // Can't set a limit if it is negative.
+                if (arg2 >= 0) {
+                    // If slice is given two arguments set limit to the position we want to slice.
+                    dynamic_cast<ExpressionFilter*>(vpOperand[0].get())->setLimit(arg2);
+                }
+            } else if (vpOperand.size() > 2) {
+                if (auto thirdArg = dynamic_cast<ExpressionConstant*>(vpOperand[2].get())) {
+                    auto thirdVal = thirdArg->getValue();
+
+                    uassert(50799,
+                            str::stream() << "Third argument to $slice must be numeric, but "
+                                          << "is of type: " << typeName(thirdVal.getType()),
+                            thirdVal.numeric());
+
+                    uassert(50800,
+                            str::stream() << "Third argument to $slice can't be represented"
+                                          << " as a 32-bit integer: " 
+                                          << thirdVal.coerceToDouble(),
+                            thirdVal.integral());
+
+                    uassert(50801,
+                            str::stream() << "Third argument to $slice must be positive: "
+                                          << thirdVal.coerceToInt(),
+                            thirdVal.coerceToInt() > 0);
+
+                    int arg3 = thirdVal.coerceToInt();
+                    if (arg2 >= 0) {
+                        // If ExpressionSlice is given three arguments set limit to 'firstArg' +
```

Comment:
Can you elaborate a little bit more here? Maybe something like `If we have three arguments, then we can stop the $filter once we have seen 'firstArg' + 'secondArg' + 1. We need to skip 'firstArg' elements, then include 'secondArg' things after that. For example, if we want to slice starting from 2 with length 3, we skip the first 2, then include 3 after that.`

Wait, I think we don't need the `+1` here? Is that right?